File: 125.Valid-Palindrome/pointers.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if len(s) <= 1:
            return True

        # O(n)
        chars = [x.lower() for x in s if x.isalnum()]

        i = 0
        j = len(s) - 1

        # O(n)
        while i < j:
            if chars[i] != chars[j]:
                return False

            i += 1
            j -= 1

        return True

File: 125.Valid-Palindrome/pointers.rs
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        if s.is_empty() {
            return true;
        }

        let mut chars = s
            .chars()
            .filter_map(|c| c.is_ascii_alphanumeric().then_some(c.to_ascii_lowercase()));

        // O(n)
        while let (Some(front), Some(back)) = (chars.next(), chars.next_back()) {
            if front != back {
                return false;
            }
        }

        true
    }
}
